feat(gux-radial-loading): improve accessibility for radial loading components

katie-bobbe-genesys opened the pull request.

daragh-king-genesys commented on src/components/stable/gux-radial-loading/gux-radial-loading.less:
```diff
@@ -58,5 +58,16 @@ gux-radial-loading {
+      margin: -1px;
+      overflow: hidden;
+      clip: rect(0, 0, 0, 0);
+      border: 0;
```
Move this css into the mixin file and name it something like gux-screenreader-only. Then you can import it here and use it in your class and we can reuse it in other components as we add screenreader only elements

katie-bobbe-genesys replied:
Good idea. Done

daragh-king-genesys commented on src/components/stable/gux-radial-progress/gux-radial-progress.tsx:
```diff
@@ -45,47 +55,55 @@ export class GuxRadialProgress {
+        <span class="gux-progress-alert" role="alert" aria-live="polite">
+          {Math.round((value / max) * 100) < 100
+            ? `${Math.round((value / max) * 100)}%`
+            : this.getI18nValue('loadingComplete')}
```
I don't think we should have special text instead of 100%. We can not guarantee that this will be used for loading.

katie-bobbe-genesys replied:
Good point. I removed the loading complete text.

daragh-king-genesys commented on src/components/stable/gux-radial-progress/gux-radial-progress.tsx:
```diff
@@ -45,47 +55,55 @@ export class GuxRadialProgress {
+              dominant-baseline="central"
+              class="gux-percentage"
+            >
+              {`${Math.round((value / max) * 100)}%`}
```
could the aria-live attribute be used here if we do not have a special 100% text

katie-bobbe-genesys replied:
I may be mistaken, but I believe the aria-live attribute that announces the percent loaded needs to be outside of the progressbar in its own span. I could not get the aria-live to work if the aria-live attribute was within the progressbar.

daragh-king-genesys reviewed (changes requested):
I think this is good for now but we should look into creating a new component that slots in a label element and a progress progress element to get better native support. If others agree we can add it to the v3 roadmap.

katie-bobbe-genesys commented on src/components/stable/gux-radial-loading/i18n/ar.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "loading": ""
```
What is the process for populating these i18n files with translations?

daragh-king-genesys replied:
You create a ticket in the LOCAL Jira board. Here is an example of one https://inindca.atlassian.net/browse/LOCAL-7185

daragh-king-genesys replied:
You create a ticket in the LOCAL Jira board. Here is an example of one https://inindca.atlassian.net/browse/LOCAL-7185

MattCheely commented:
> I think this is good for now but we should look into creating a new component that slots in a label element and a progress progress element to get better native support. If others agree we can add it to the v3 roadmap.

I would actually look at using the native `progressbar` element in this PR, or the `progressbar` aria role, and maybe adding a mechanism to link to the target loading item.  That could prevent the need for a future breaking API change.

@katie-bobbe-genesys was there a particular reference you used in figuring out how to approach this?  I found some [MDN guidelines](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_progressbar_role) that suggested using a link like `described-by` to relate the loader to the loading item, which might be less intrusive than making the loaders live if there are multiple on a page.

katie-bobbe-genesys commented:
I changed the implementation for adding accessibility to these components in response to some of the above comments. I used the `progressbar` aria-role and now require authors to provide screenreader text for the component. I added a console error similar to the one used with the icons if the author does not add the screenreader-text property to the component. The screenreader-text property will take away the need to have translation files for the "Loading" string and will provide authors with more flexibility with the component. I am still exploring a way to include `described-by` to provide a link between the loader and the loading element. Let me know if anyone has any ideas/advice for how I might want to go about this.

katie-bobbe-genesys commented on src/components/stable/gux-radial-progress/gux-radial-progress.tsx:
```diff
@@ -27,75 +27,106 @@ export class GuxRadialProgress {
+   * Optional value to specify if the component is being used to represent a static meter element.
+   */
+  @Prop()
+  meter: boolean = false;
```
I was experimenting with adding a `meter` attribute. This will only change the role from `progressbar` to `meter`. If someone wants to use this `gux-radial-progress` component to show a static value instead of showing progress, the `progressbar` role would be confusing to screen reader users. An example of someone using this component as a meter is here: https://inindca.atlassian.net/browse/COMUI-422

MattCheely replied:
I think the usage of a circle representing a percentage or similar value is really a data visualization problem.  It's unfortunate that this component is getting used that way now, but if we put meter support in here, it makes it harder to get folks to transition away to an eventual dataviz component, and I do think we want separate components so that we can evolve the presentation of percentages without impacting loader behavior.

I'd rather leave this out for now and aim for a long-term fix in the data visualization work.

daragh-king-genesys commented on src/components/stable/gux-radial-progress/gux-radial-progress.less:
```diff
@@ -29,6 +29,17 @@ gux-radial-progress {
+    margin: -1px;
+    overflow: hidden;
+    clip: rect(0, 0, 0, 0);
+    border: 0;
```
This can use the `.gux-sr-only` mixin

daragh-king-genesys commented on src/components/stable/gux-page-loading-spinner/gux-page-loading-spinner.tsx:
```diff
@@ -10,11 +10,30 @@ export class GuxPageLoadingSpinner {
+   * Required localized text to provide screen reader accessible label for the component
+   */
+  @Prop()
+  screenreaderText: string = '';
```
Not sure if I am backtracking here but maybe `gux-page-loading-spinner` should have default translated text.
I know the progress component cant but the loading ones probably can.

daragh-king-genesys commented on src/components/stable/gux-radial-loading/tests/gux-radial-progress.spec.ts:
```diff
@@ -20,10 +20,10 @@ describe('gux-radial-loading', () => {
+      '<gux-radial-loading screenreader-text="Loading"></gux-radial-loading>',
+      '<gux-radial-loading context="modal" screenreader-text="Loading"></gux-radial-loading>',
+      '<gux-radial-loading context="full-page" screenreader-text="Loading"></gux-radial-loading>',
+      '<gux-radial-loading context="input" screenreader-text="Loading"></gux-radial-loading>'
```
You could add tests here for when no screenreader test is added

daragh-king-genesys commented on src/components/stable/gux-radial-progress/tests/gux-radial-progress.spec.ts:
```diff
@@ -20,18 +20,18 @@ describe('gux-radial-progress', () => {
+      '<gux-radial-progress value="test" max="100" screenreader-text="Loading"></gux-radial-progress>',
+      '<gux-radial-progress value="-123" max="100" screenreader-text="Loading"></gux-radial-progress>',
+      '<gux-radial-progress value="200" max="100" screenreader-text="Loading"></gux-radial-progress>',
+      '<gux-radial-progress value="10" max="test" screenreader-text="Loading"></gux-radial-progress>'
```
I don't think loading is a good example of screenreader text in this component. Maybe `Uploading file` or`Generating model`
